File: sequencer/data_store.py
def singleton(cls):
    instances = {}

    def get_instance(*args, **kwargs):
        if cls not in instances:
            instances[cls] = cls(*args, **kwargs)
        return instances[cls]

    return get_instance


@singleton
class MenusContainer:
    def __init__(self):
        self._data = (
            ["STEP", "OFFSET", "LIMIT", "NOTE"],
            ["BPM", "GATE", "VELOCITY", "EACH"],
            ["SWING", "SIGN", "EMPTY", "EMPTY"],
            ["EMPTY", "EMPTY", "EMPTY", "EMPTY"],
        )  # Each page has four visible parameters

    def get_value(self, index):
        return self._data[index]


@singleton
class DisplayDataContainer:
    def __init__(self):
        self._data = {
            "page": 1,
            "menu_position": 1,
        }  # Current state of menu and option parameter

    def get_value(self, key):
        return self._data.get(key)

    def set_value(self, key, value):
        self._data[key] = value

File: src/data_store.cpp
#include "data_store.hpp"

const char *menu::names[4][6] = {{"STEP", "OFF", "LIMIT", "NOTE", "BPM"},
                                 {"CHAN", "GATE", "VELO", "EACH", "SWING"},
                                 {"SIGN", nullptr, nullptr, nullptr, nullptr},
                                 {nullptr, nullptr, nullptr, nullptr, nullptr}};

int menu::menu_position = 1;
int menu::submenu_position = 1;
int menu::channel = 1;

Channel channel::channels[8] = {Channel(1), Channel(2), Channel(3), Channel(4),
                       Channel(5), Channel(6), Channel(7), Channel(8)};
